File: app/src/main/java/com/example/tradeup/data/source/remote/FirebaseChatSource.java
package com.example.tradeup.data.source.remote;

import android.util.Log;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.example.tradeup.core.utils.Callback;
import com.example.tradeup.data.model.Chat;    // Model Chat (Java)
import com.example.tradeup.data.model.Message;  // Model Message (Java)
import com.google.firebase.firestore.CollectionReference;
import com.google.firebase.firestore.DocumentReference;
import com.google.firebase.firestore.DocumentSnapshot;
import com.google.firebase.firestore.FieldValue;
import com.google.firebase.firestore.FirebaseFirestore;
import com.google.firebase.firestore.ListenerRegistration;
import com.google.firebase.firestore.Query;
import com.google.firebase.firestore.WriteBatch;


import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import javax.inject.Inject;

public class FirebaseChatSource {

    private final FirebaseFirestore firestore;
    private final CollectionReference chatsCollection;

    @Inject
    public FirebaseChatSource(FirebaseFirestore firestore) {
        this.firestore = firestore;
        this.chatsCollection = firestore.collection("chats");
    }

    public void getOrCreateChat(List<String> participantIds, @Nullable String relatedItemId, Callback<String> callback) {
        if (participantIds == null || participantIds.size() != 2) {
            callback.onFailure(new IllegalArgumentException("Chat must have exactly 2 participants."));
            return;
        }

        List<String> sortedIds = new ArrayList<>(participantIds);
        Collections.sort(sortedIds);
        // Cân nhắc việc có nên bao gồm relatedItemId trong chatId hay không.
        // Nếu mỗi sản phẩm có 1 chat riêng giữa 2 người:
        // String chatId = sortedIds.get(0) + "_" + sortedIds.get(1) + (relatedItemId != null ? "_" + relatedItemId : "");
        // Nếu chỉ có 1 chat duy nhất giữa 2 người:
        String chatId = sortedIds.get(0) + "_" + sortedIds.get(1);


        DocumentReference chatDocRef = chatsCollection.document(chatId);
        chatDocRef.get().addOnCompleteListener(task -> {
            if (task.isSuccessful()) {
                if (task.getResult() != null && task.getResult().exists()) {
                    callback.onSuccess(chatId); // Chat đã tồn tại
                } else {
                    // Tạo chat mới
                    Chat newChat = new Chat();

                    // KHÔNG CẦN DÒNG NÀY NỮA, VÌ @DocumentId SẼ TỰ LÀM:
                    // newChat.setChatId(chatId);

                    newChat.setParticipants(sortedIds);
                    if (relatedItemId != null) {
                        newChat.setRelatedItemId(relatedItemId);
                    }
                    // Các trường khác như participantInfo, unreadCount...

                    // Khi set(), Firestore sẽ tự động điền ID document vào trường có @DocumentId
                    chatDocRef.set(newChat)
                            .addOnSuccessListener(aVoid -> callback.onSuccess(chatId))
                            .addOnFailureListener(callback::onFailure);
                }
            } else {
                callback.onFailure(task.getException());
            }
        });
    }

    public ListenerRegistration getChatList(String userId, Callback<List<Chat>> callback) {
        return chatsCollection
                .whereArrayContains("participants", userId)
                .orderBy("lastMessageTimestamp", Query.Direction.DESCENDING)
                .addSnapshotListener((snapshots, e) -> {
                    if (e != null) {
                        callback.onFailure(e);
                        return;
                    }
                    if (snapshots != null) {
                        List<Chat> chats = new ArrayList<>();
                        for (DocumentSnapshot doc : snapshots.getDocuments()) {
                            Chat chat = doc.toObject(Chat.class);
                            if (chat != null) {
                                // Firestore không tự gán ID document vào object khi dùng toObject với listener
                                // Nếu Chat.java của bạn có setter cho chatId và field chatId
                                // chat.setChatId(doc.getId()); // Gán ID document thủ công
                                chats.add(chat);
                            }
                        }
                        callback.onSuccess(chats);
                    }
                });
    }

    public ListenerRegistration getMessages(String chatId, Callback<List<Message>> callback) {
        return chatsCollection.document(chatId).collection("messages")
                .orderBy("timestamp", Query.Direction.ASCENDING)
                .addSnapshotListener((snapshots, e) -> {
                    if (e != null) {
                        callback.onFailure(e);
                        return;
                    }
                    if (snapshots != null) {
                        List<Message> messages = new ArrayList<>();
                        for (DocumentSnapshot doc : snapshots.getDocuments()) {
                            Message message = doc.toObject(Message.class);
                            if (message != null) {
                                // message.setMessageId(doc.getId()); // Gán ID nếu cần
                                messages.add(message);
                            }
                        }
                        callback.onSuccess(messages);
                    }
                });
    }

    public void sendMessage(String chatId, Message message, Callback<Void> callback) {
        CollectionReference messagesSubCollection = chatsCollection.document(chatId).collection("messages");

        WriteBatch batch = firestore.batch();

        // 1. Thêm tin nhắn mới vào subcollection
        // Dùng trực tiếp object Message vì nó đã có @ServerTimestamp cho trường timestamp
        DocumentReference newMessageRef = messagesSubCollection.document();
        batch.set(newMessageRef, message);

        // 2. Cập nhật thông tin trong document chat chính
        DocumentReference chatDocRef = chatsCollection.document(chatId);
        Map<String, Object> chatUpdates = new HashMap<>();

        String lastMessageText = message.getText();
        if (lastMessageText == null && message.getImageUrl() != null) {
            lastMessageText = "Image";
        } else if (lastMessageText == null && message.getOfferDetails() != null) {
            lastMessageText = "Offer";
        }

        chatUpdates.put("lastMessageText", lastMessageText);
        chatUpdates.put("lastMessageTimestamp", FieldValue.serverTimestamp());
        chatUpdates.put("lastMessageSenderId", message.getSenderId());
        chatUpdates.put("updatedAt", FieldValue.serverTimestamp());

        // === PHẦN SỬA LỖI & THÊM MỚI QUAN TRỌNG NHẤT ===
        // Tăng unreadCount của người nhận (receiver) lên 1
        if (message.getReceiverId() != null) {
            // Dùng dot notation để cập nhật một trường bên trong một map
            // Ví dụ: "unreadCount.user_id_B"
            chatUpdates.put("unreadCount." + message.getReceiverId(), FieldValue.increment(1));
        }
        // === KẾT THÚC PHẦN SỬA LỖI ===

        batch.update(chatDocRef, chatUpdates);

        batch.commit()
                .addOnSuccessListener(aVoid -> {
                    Log.d("FirebaseChatSource", "sendMessage batch commit successful.");
                    callback.onSuccess(null);
                })
                .addOnFailureListener(e -> {
                    Log.e("FirebaseChatSource", "sendMessage batch commit failed.", e);
                    callback.onFailure(e);
                });
    }

    public void markMessagesAsRead(String chatId, String userId, Callback<Void> callback) {
        // Logic này có thể phức tạp:
        // 1. Cập nhật trường unreadCount.<userId> = 0 trong Chat document.
        // 2. (Tùy chọn) Query các tin nhắn chưa đọc của user đó trong subcollection messages và update isRead = true.
        // Bước 1 thường đủ cho việc hiển thị số lượng tin nhắn chưa đọc.

        chatsCollection.document(chatId)
                .update("unreadCount." + userId, 0) // Ví dụ: "unreadCount.userId123"
                .addOnSuccessListener(aVoid -> callback.onSuccess(null))
                .addOnFailureListener(callback::onFailure);
    }
}

File: app/src/main/java/com/example/tradeup/data/source/remote/FirebaseChatSource.kt
package com.example.tradeup.data.source.remote // Đã đổi

import com.google.firebase.firestore.FirebaseFirestore
import com.google.firebase.firestore.Query
import com.google.firebase.firestore.ktx.snapshots
import com.google.firebase.firestore.ktx.toObject
import com.example.tradeup.data.model.Chat // Đã đổi
import com.example.tradeup.data.model.Message // Đã đổi
import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.map
import kotlinx.coroutines.tasks.await
import javax.inject.Inject

@ExperimentalCoroutinesApi
class FirebaseChatSource @Inject constructor(private val firestore: FirebaseFirestore) {

    private val chatsCollection = firestore.collection("chats")

    suspend fun getOrCreateChat(participantIds: List<String>, relatedItemId: String? = null): Result<String> {
        if (participantIds.size != 2) return Result.failure(IllegalArgumentException("Chat must have 2 participants"))
        val sortedIds = participantIds.sorted()
        val chatId = "${sortedIds[0]}_${sortedIds[1]}"

        return try {
            val chatDoc = chatsCollection.document(chatId)
            val snapshot = chatDoc.get().await()
            if (snapshot.exists()) {
                Result.success(chatId)
            } else {
                val newChat = Chat(
                    chatId = chatId,
                    participants = sortedIds,
                    relatedItemId = relatedItemId
                )
                chatDoc.set(newChat).await()
                Result.success(chatId)
            }
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    fun getChatList(userId: String): Flow<Result<List<Chat>>> {
        return chatsCollection
            .whereArrayContains("participants", userId)
            .orderBy("lastMessageTimestamp", Query.Direction.DESCENDING)
            .snapshots()
            .map { querySnapshot ->
                try {
                    val chats = querySnapshot.documents.mapNotNull { it.toObject<Chat>() }
                    Result.success(chats)
                } catch (e: Exception) {
                    Result.failure(e)
                }
            }
    }

    fun getMessages(chatId: String): Flow<Result<List<Message>>> {
        return chatsCollection.document(chatId).collection("messages")
            .orderBy("timestamp", Query.Direction.ASCENDING)
            .snapshots()
            .map { querySnapshot ->
                try {
                    val messages = querySnapshot.documents.mapNotNull { it.toObject<Message>() }
                    Result.success(messages)
                } catch (e: Exception) {
                    Result.failure(e)
                }
            }
    }

    suspend fun sendMessage(chatId: String, message: Message): Result<Unit> {
        return try {
            chatsCollection.document(chatId).collection("messages").add(message).await()
            chatsCollection.document(chatId).update(
                mapOf(
                    "lastMessageText" to (message.text ?: (if(message.imageUrl != null) "Image" else "Offer")),
                    "lastMessageTimestamp" to message.timestamp,
                    "lastMessageSenderId" to message.senderId,
                    "updatedAt" to com.google.firebase.firestore.FieldValue.serverTimestamp()
                )
            ).await()
            Result.success(Unit)
        } catch (e: Exception) {
            Result.failure(e)
        }
    }
    suspend fun markMessagesAsRead(chatId: String, userId: String): Result<Unit> {
        return try {
            chatsCollection.document(chatId).update("unreadCount.$userId", 0).await()
            Result.success(Unit)
        } catch (e: Exception) {
            Result.failure(e)
        }
    }
}